### PR TITLE
add permit functionality

### PR DESCRIPTION
**Pull request by shbhakta**

### CODE REVIEW

File: contracts/tokenbridge/libraries/gateway/GatewayRouter.sol [hunk 1]
```diff
@@ -144,4 +157,132 @@ abstract contract GatewayRouter is TokenGateway, IGatewayRouter {
...
+     * @param _to Account to be credited with the tokens in the L2 (can be the user's L2 account or a contract), not subject to L2 aliasing
+                  This account, or its L2 alias if it have code in L1, will also be able to cancel the retryable ticket and receive callvalue refund
+     * @param _amount Token Amount
+     * @param _maxGas Max gas deducted from user's L2 balance to cover L2 execution
+     * @param _gasPriceBid Gas price for L2 execution
+     * @param _data encoded data from router and user
+     * @param _permitData signature and deadline params of permit
+     * @return res abi encoded inbox sequence number
+    */
+    function outboundTransferCustomRefundWithPermit(
```
Comment (gzeoneth) [inline contracts/tokenbridge/libraries/gateway/GatewayRouter.sol:237]: The `CustomRefund` variant should be implemented on L1 only, like how we have [this in IL1GatewayRouter](https://github.com/OffchainLabs/token-bridge-contracts/blob/0c4ffc32291ecca243f82723f1dee72ff347aa01/contracts/tokenbridge/ethereum/gateway/IL1GatewayRouter.sol#L48)

Comment (gzeoneth) [reply contracts/tokenbridge/libraries/gateway/GatewayRouter.sol:237]: So I think the plan would be to have an internal function that do permit stuff in `GatewayRouter`, and then `outboundTransferWithPermit` in `GatewayRouter` and `outboundTransferCustomRefundWithPermit` in `L1GatewayRouter`

File: contracts/tokenbridge/libraries/gateway/GatewayRouter.sol [hunk 2]
```diff
@@ -144,4 +157,132 @@ abstract contract GatewayRouter is TokenGateway, IGatewayRouter {
...
+                _permitData.deadline,
+                true,
+                _permitData.v,
+                _permitData.r,
+                _permitData.s
+            );
+        }
+
+        return
+            IL1ArbitrumGateway(gateway).outboundTransferCustomRefund{ value: msg.value }(
```
Comment (gzeoneth) [inline contracts/tokenbridge/libraries/gateway/GatewayRouter.sol:278]: Should not use `IL1ArbitrumGateway` in `GatewayRouter` since it should support both L1 and L2

File: contracts/tokenbridge/libraries/IDaiLikePermit.sol [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+interface IDaiLikePermit {
+    function permit(
+        address holder,
+        address spender,
+        uint256 nonce,
+        uint256 expiry,
+        bool allowed,
+        uint8 v,
```
Comment (gzeoneth) [inline contracts/tokenbridge/libraries/IDaiLikePermit.sol:20]: Can reorder these variables to use less calldata/storage slot

**Review by gzeoneth — CHANGES_REQUESTED** [review]
Looking good, it's much cleaner now. Added some more suggestions.

File: contracts/tokenbridge/libraries/gateway/GatewayRouter.sol [hunk 4]
```diff
@@ -36,6 +40,15 @@ abstract contract GatewayRouter is TokenGateway, IGatewayRouter {
...
     address public override defaultGateway;
 
+    struct PermitData {
+        uint256 deadline;
+        uint256 nonce;
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        bool isStandardImpl;
+    }
```
Comment (gzeoneth) [inline contracts/tokenbridge/libraries/gateway/GatewayRouter.sol:50]: can move `isStandardImpl` between `nonce` and `v`

**Review by gzeoneth — CHANGES_REQUESTED** [review]

File: contracts/tokenbridge/ethereum/gateway/L1GatewayRouter.sol [hunk 5]
```diff
@@ -304,6 +305,62 @@ contract L1GatewayRouter is
...
+            super.outboundTransferWithPermit(
+                _l1Token,
+                _to,
+                _amount,
+                _maxGas,
+                _gasPriceBid,
+                _data,
+                _permitData
+            );
+    }
```
Comment (gzeoneth) [inline contracts/tokenbridge/ethereum/gateway/L1GatewayRouter.sol:327]: I think it will still compile without this function override.

File: contracts/tokenbridge/libraries/gateway/GatewayRouter.sol [hunk 6]
```diff
@@ -17,12 +17,16 @@
  */
 
 pragma solidity ^0.6.11;
+pragma experimental ABIEncoderV2;
 
 import "arb-bridge-eth/contracts/libraries/ProxyUtil.sol";
 import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/drafts/ERC20Permit.sol";
 import "./TokenGateway.sol";
+import "../../ethereum/gateway/IL1ArbitrumGateway.sol";
```
Comment (gzeoneth) [inline contracts/tokenbridge/libraries/gateway/GatewayRouter.sol:26]: can remove

File: yarn.lock [hunk 7]
```diff
@@ -4,7 +4,7 @@
 
 "@babel/code-frame@7.12.11":
   version "7.12.11"
-  resolved "https://registry.yarnpkg.com/@babel/code-frame/-/code-frame-7.12.11.tgz#f4ad435aa263db935b8f10f2c552d23fb716a63f"
+  resolved "https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.12.11.tgz"
```
Comment (gzeoneth) [inline yarn.lock:7]: I think we can revert the change to yarn.lock since there are no new package / version bump, not a big deal

**Review by gzeoneth — COMMENTED** [review]
added some final suggestion, I think its pretty good @fredlacs wdyt?

**Review by gzeoneth — APPROVED** [review]

File: contracts/tokenbridge/libraries/gateway/GatewayRouter.sol [hunk 8]
```diff
@@ -144,4 +157,59 @@ abstract contract GatewayRouter is TokenGateway, IGatewayRouter {
...
+     * @param _token L1 address of ERC20
+     * @param _to Account to be credited with the tokens in the L2 (can be the user's L2 account or a contract), not subject to L2 aliasing
+                  This account, or its L2 alias if it have code in L1, will also be able to cancel the retryable ticket and receive callvalue refund
+     * @param _amount Token Amount
+     * @param _maxGas Max gas deducted from user's L2 balance to cover L2 execution
+     * @param _gasPriceBid Gas price for L2 execution
+     * @param _data encoded data from router and user
+     * @param _permitData signature and deadline params of permit
+    */
+    function outboundTransferWithPermit(
```
Comment (fredlacs) [inline contracts/tokenbridge/libraries/gateway/GatewayRouter.sol:173]: would be nice to add this to `IGatewayRouter`

File: contracts/tokenbridge/libraries/gateway/GatewayRouter.sol [hunk 9]
```diff
@@ -144,4 +157,59 @@ abstract contract GatewayRouter is TokenGateway, IGatewayRouter {
...
+        return outboundTransfer(_token, _to, _amount, _maxGas, _gasPriceBid, _data);
+    }
+
+    function callPermit(
+        address _token,
+        uint256 _amount,
+        PermitData calldata _permitData
+    ) internal virtual {
+        address gateway = getGateway(_token);
+        if (_permitData.isStandardImpl) {
```
Comment (fredlacs) [inline contracts/tokenbridge/libraries/gateway/GatewayRouter.sol:192]: I think we could implement a separate entrypoint function instead of `bool isStandardImpl`
something like `outboundTransferWithEip2612Permit` and `outboundTransferWithDaiPermit`
would save us a bit of gas

File: contracts/tokenbridge/libraries/gateway/GatewayRouter.sol [hunk 10]
```diff
@@ -36,6 +40,15 @@ abstract contract GatewayRouter is TokenGateway, IGatewayRouter {
...
     address public override defaultGateway;
 
+    struct PermitData {
+        uint256 deadline;
+        uint256 nonce;
+        bool isStandardImpl;
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+    }
```
Comment (fredlacs) [inline contracts/tokenbridge/libraries/gateway/GatewayRouter.sol:50]: we probably want to declare this in an interface file instead

File: contracts/tokenbridge/libraries/gateway/GatewayRouter.sol [hunk 11]
```diff
@@ -144,4 +157,59 @@ abstract contract GatewayRouter is TokenGateway, IGatewayRouter {
...
+
+    function callPermit(
+        address _token,
+        uint256 _amount,
+        PermitData calldata _permitData
+    ) internal virtual {
+        address gateway = getGateway(_token);
+        if (_permitData.isStandardImpl) {
+            ERC20Permit(_token).permit(
```
Comment (fredlacs) [inline contracts/tokenbridge/libraries/gateway/GatewayRouter.sol:193]: would be good to add a comment here explaining why its safe from reentrancy

File: contracts/tokenbridge/libraries/gateway/GatewayRouter.sol [hunk 12]
```diff
@@ -170,46 +161,61 @@ abstract contract GatewayRouter is TokenGateway, IGatewayRouter {
...
     */
-    function outboundTransferWithPermit(
+    function outboundTransferWithEip2612Permit(
         address _token,
         address _to,
         uint256 _amount,
         uint256 _maxGas,
         uint256 _gasPriceBid,
         bytes calldata _data,
         PermitData calldata _permitData
```
Comment (fredlacs) [inline contracts/tokenbridge/libraries/gateway/GatewayRouter.sol:171]: Lets make a different struct? we don't need deadline here

**Review by fredlacs — APPROVED** [review]
LGTM

File: contracts/tokenbridge/libraries/gateway/GatewayRouter.sol [hunk 13]
```diff
@@ -144,4 +147,77 @@ abstract contract GatewayRouter is TokenGateway, IGatewayRouter {
...
+    function outboundTransferWithEip2612Permit(
+        address _token,
+        address _to,
+        uint256 _amount,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        bytes calldata _data,
+        PermitData calldata _permitData
+    ) public payable virtual override returns (bytes memory) {
+        address gateway = getGateway(_token);
```
Comment (gzeoneth) [inline contracts/tokenbridge/libraries/gateway/GatewayRouter.sol:173]: Actually we can also have the gateway as 1 of the external function parameter, since the UI need to figure out which gatway to sign permit anyway.
